com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/detect/DeployedMtaEnvDetector.java

```diff
@@ -0,0 +1,121 @@
+package com.sap.cloud.lm.sl.cf.core.cf.detect;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.cloudfoundry.client.lib.CloudControllerClient;
+import org.cloudfoundry.client.lib.domain.CloudApplication;
+import org.springframework.stereotype.Component;
+
+import com.sap.cloud.lm.sl.cf.core.Constants;
+import com.sap.cloud.lm.sl.cf.core.cf.metadata.ImmutableMtaMetadata;
+import com.sap.cloud.lm.sl.cf.core.cf.metadata.MtaMetadata;
+import com.sap.cloud.lm.sl.cf.core.cf.metadata.processor.EnvMtaMetadataParser;
+import com.sap.cloud.lm.sl.cf.core.model.DeployedMta;
+import com.sap.cloud.lm.sl.cf.core.model.DeployedMtaModule;
+import com.sap.cloud.lm.sl.cf.core.model.DeployedMtaResource;
+import com.sap.cloud.lm.sl.cf.core.model.ImmutableDeployedMta;
+import com.sap.cloud.lm.sl.mta.model.Version;
+
+//TODO delete this class and its usages after the CF metadata becomes the go to metadata approach
+//A release note should be already present explaining that the migration (at least one mta redeploy) is mandatory
+@Component
+public class DeployedMtaEnvDetector {
+
+    @Inject
+    private EnvMtaMetadataParser envMtaMetadataParser;
+
+    public List<DeployedMta> getAllDeployedMtas(CloudControllerClient client) {
+        Map<String, List<CloudApplication>> applicationsByMtaId = getApplicationsByMtaId(client);
+        return applicationsByMtaId.values()
+                                  .stream()
+                                  .map(this::toDeployedMta)
+                                  .collect(Collectors.toList());
+    }
+
+    private Map<String, List<CloudApplication>> getApplicationsByMtaId(CloudControllerClient client) {
+        Map<String, List<CloudApplication>> applicationsByMtaId = new HashMap<>();
+        List<CloudApplication> applications = getApplicationsWithEnvMetadata(client);
+        for (CloudApplication application : applications) {
```

Comment:
This can be substituted with 
`applications.stream().collect(groupingBy(app -> parseMetadata(app).getId())`

Reply:
This used to be groupingBy but got changed recently as a discussion with Encho led us to agree that we wanted to avoid having the complexity of 'groupingBy' in the code. What do you think?

Reply:
I half agree. Using grouping by in complex scenarios, like having to modify the downstream (eg. `groupingBy(<some property>, mapping()/collectingAndThen()/toSet()...), is worse than using a regular loop. But this is a standard scenario. My opinion is that all the extra lines and extra operations on the map are more complex than using `stream().collect(groupingBy(func()))`